proposals/2022-02-08-Websockets.md

```diff
@@ -0,0 +1,230 @@
+      operationId: "sampleSocket"
+      publish:
+        description: "Short description about the messages you can push towards the server"
+        content:
+          application/json:
+            schema:
+              oneOf:
+                - $ref: "#/components/schemas/CreateUserSchema"
+                - $ref: "#/components/schemas/DeleteUserSchema"
+                - $ref: "#/components/schemas/LikePostSchema"
+      subscribe:
+        description: "Short description about the messages you can receive from the server"
+        content:
+          application/json:
+            schema:
+              oneOf:
+                - $ref: "#/components/schemas/UserCreatedSchema"
+                - $ref: "#/components/schemas/UserDeletedSchema"
+                - $ref: "#/components/schemas/PostLikedSchema"
+```
+
+## Detailed design
+
+The [WebSocket spec](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) is currently very simple. From our point of view it allows:
+
+1. Connecting to an endpoint using a URL
+2. Sending messages in either binary or text format
+3. Receiving messages in either binary or text format
+
+This proposal aims to allow OpenAPI users to describe exactly the above mentioned points:
+
+1. What URL can I connect to, to talk to this WebSocket?
+2. What's the format and shape of data I'm allowed to send
+3. What's the format and shape of data I can expect to receive
+
+**Important**: This OpenAPI spec proposal doesn't concern itself with any frameworks, or any higher level abstractions people usually build on top of event driven messaging protocols, like topics/subjects/channels, implied message serialization, etc. It simply describes the above mentioned 3 points, and any further abstractions can be built on top of this, but not part of the core schema.
+
+### Extension of the current OpenAPI spec
+
+- Add a `WebSocket Payload Object` type (very similar to [Request Body Object](https://swagger.io/specification/#request-body-object), except the `required` field - which makes no sense here). It has the following fields:
+  - `description`: `string` - A brief description of the payload.
+  - `content`: `Map[string, `[Media Type Object](https://swagger.io/specification/#media-type-object)`]` The content of the payload.
+- Add an optional field in the [Components Object](https://swagger.io/specification/#components-object) type, so payloads can be reused:
+  - `webSocketPayloads`: `Map[string, WebSocket Payload Object]` - Reuseable payloads
+- Add 2 optional fields on the [Operation Object](https://swagger.io/specification/#operation-object) type:
+  - `publish`: [Reference Object](https://swagger.io/specification/#reference-object) | `WebSocket Payload Object` - describes the message(s) the client can publish/push to the server.
+  - `subscribe`: [Reference Object](https://swagger.io/specification/#reference-object) | `WebSocket Payload Object` - describes the message(s) the client can receive from the server.
```

Comment:
If both fields can be optional, how can a tool that reads an OpenAPI document identify a web socket endpoint with certainty? If the OpenAPI document author did not supply either field, the endpoint is unrecognisable from a standard endpoint.

What should happen if those fields are added to an operation other than `get`? Is that possible?

Reply:
1. I'd say if either fields `publish` or `subscribe` is present, then it's a socket enabled path. If neither is supplied, it should fall back to be recognized as regular http path. Are you suggesting something more explicit (extra boolean field maybe)?

2. I'd say error. Similar situation, as to what should happen if a `GET` request has a `requestBody`. 

What do you think?

Reply:
OK, that would work. If anyone needs more, they can always create a proposal for it!